### ISSUE TITLE
support apache arrow client

### ISSUE THREAD

**Issue by dysn** [trinodb/trino#24586]
Arrow is primarily an in memory columnar data format, but defines ways for applications to exchange such data efficiently. Many common data processing/analysis tools support working directly on arrow data efficiently (eg polars, duckdb)

At our company, we are seeing a pattern emerge where data scientists and analysts will first generate modest amounts of data (100s of mbs to 10s of gbs) with heavy trino queries, and then load that into something like polars or duckdb for further analysis. We would like to better support this pattern. One current pain point is that users must convert the row oriented trino data into arrow data. Arrow provides utilities for this, but that is not so efficient.

There are a few ways trino could return arrow data to users. One would be by implementing the arrow flight (and higher level arrow flight sql) grpc protocol. This seems like it could be a pretty big lift, and likely would require some extra up front design. It may not be possible to support an acceptable subset of trinos features on the existing arrow flight protocol. In particular, fetching data in arrow flight is a grpc endpoint, which doesnt fit well with trino's spooling protocol

An easier way might be to implement a java adbc connector for trino. ADBC is just a java interface and doesnt require any specific protocol. There are two ways we could implement and adbc connector. We could modify the direct protocol to return data in the arrow wire format, or we could enable trino to spool data in the arrow ipc format. Of course, these aren't mutually exclusive.

Is an arrow based trino client desirable? If so, how should that be implemented?

**Comment by nineinchnick** [comment]
Duplicate of #18038, you could comment there if you have a different and/or more specific use case.

**Comment by dysn** [comment]
i think this is not exactly a duplicate, but i can move discussion there if you prefer

this ticket is more generally about having some kind of arrow based client for trino, whereas 18038 is specifically about arrow flight. its also not clear to me if 18038 is about creating an arrow flight connector for trino so it can fetch data from an arrow flight server, or implementing the arrow flight endpoints in trino so it can serve as an arrow flight server

**Comment by nineinchnick** [comment]
Thanks, I see the distinction now. It might be actually a much smaller effort to use Arrow instead of JSON as the transport format, than implementing a client compatible with Arrow Flight SQL.

I think the biggest issue with Arrow is how to handle Trino data types that don't map to Arrow types. This would require special logic on the client side - we should enumerate such types and see if that would not invalidate the whole point of using Arrow, that is, if it would be simpler to convert all data on the client side.

**Comment by dysn** [comment]
i agree that no matter what protocol is used, type conversion will be tricky. i took a pass at converting trino types to arrow types, and mostly they map well. the gaps are around (unsurprisingly) time and timestamp types. arrow only supports times up to nanosecond precision, and only supports times with timezones scoped to a file or block of data, not to the row like trino does

i handled these in my draft implementation by using arrow extension types. arrow extension types are a way to mark some arrow physical type as representing nonstandard logical type. in this case, i encoded picosecond precision times as a struct with a nano timestamp and pico adjustment, and times with timezone as a struct with a time and zone offset.

sure, i can make a draft PR. note, its very much a draft, untested and in progress

**Comment by dysn** [comment]
created https://github.com/trinodb/trino/pull/24587 with the initial draft

**Comment by shohamyamin** [comment]
@dysn maybe it is worth checking if there is a chance to make arrow supporting this problematic types in arrow itself?

**Comment by dysn** [comment]
thats a good point, since migrating from extension types to natively supported arrow types would be a breaking change for clients. on the other hand, arrow is a big project, so adding new types may be slow

would you prefer that i open an issue with arrow? or would it be better for a core trino contributor to start that discussion? FWIW, i think opening up that line of communication is worthwhile, especially if trino ever wants to support arrow flight

**Comment by shohamyamin** [comment]
I am not a maintainer at Trino or something like that but I think that it is worth opening an issue at arrow to see if adding types can be acceptable
At the same time having this conversation at Trino and start the work here. If there will be progress from arrow side that will be great and make things simple.

But it is worth to hear more from the Trino maintainers as they have more experience with integrating with other open-source projects and know more about Trino core

**Comment by dysn** [comment]
i looked into arrow flight and flight sql. much of the protocol translates well to trino, with a couple exceptions

first, flight handles fetching actual result data as a grpc call, which doesnt map to trino spooling protocol with the default STORAGE retrieval mode. I am optimistic that the arrow team might would be open to adding regular http based data fetching.

more difficult is session management. flight uses a session id cookie, but that implies that the server will track session properties and prepared statements. its certainly possible to track those in trino, but it would come with more questions and design i think. its also possible to continue storing session info with headers using a custom grpc interceptor, but that kind of partly defeats the goal of being able to query trino with a regular flight sql client

**Comment by dysn** [comment]
@wendigo suggested in slack that we could implement an arrow flight proxy server to avoid having to mix arrow dependencies and a grpc server into the existing trino server

**Comment by devozerov** [comment]
We have a fully-fledged Arrow Flight SQL implementation that already works in production clusters. Some highlights:

- Implemented over the **direct** protocol, because it was implemented before the spooling protocol appeared. Implementation on top of the spooling protocol might be beneficial for some cases but would be much more involved in terms of fault tolerance. In our experience, integration capabilities are more important to many users compared to raw theoretical performance
- Arrow, GRPC, and Jetty dependencies are well-isolated into a separate module, where they are shaded and relocated. The module is then linked into runtime via reflection. We did this to avoid dependency hell and to have a workaround should some complicated security bug appear. Potentially, we can further isolate this via an additional classloader (or reuse `PluginClassloader`)
- Supports plain data types. No complex type support for now
- Server-side session management with periodic session cleanup

We are ready to donate our implementation. However, as we all know, communication in open-source projects might be complicated. A number of communications that we initiated on various issues ended up with silence. I completely understand why this happens, worked many years for a key vendor behind an ASF project, and external contributions were always painful to the team due to a very limited time budget. However, our team is very busy also. So to initiate the donation, we need a commitment from the Starburst team and possibly other major players that sufficient time of engineers that ultimately make the merge decision will be reserved for that effort. There are approximately 10K lines of code, of which ~1K is pretty dense with a lot of concurrency, futures, etc. 

Also, since each team has a unique coding culture, and there is always legacy code, or code that someone wanted to improve but never had time, etc, it would be important to have a practical process that avoids nitpicking and unreasonably high bars. Otherwise, we are risking getting stuck in the middle. For instance, we have some ugly copy-paste of around ~200 LOC from tino-client that we would like to avoid but never have time to polish. However, the code works and brings value, and this is what matters most. During the donation, we are unlikely to have time to polish everything. So our expectations are that (1) we provide a working solution, align it with Trino guidelines in advance, polish everything we can at the moment, provide extensive docs, fix all bugs observed during the review, but (2) the other side doesn't demand an ideal solution and is ready for compromises in the form of TODOs, GitHub issues, migrating our current issues, etc.

I'm sorry if this sounds a bit offensive; that is not my goal. I just want to be candid in advance. Donations are great for an outside observer. But for the donator and receiver, they are also a pain, sometimes considerable. I saw too many failed donation attempts that either stopped in the middle or proceeded, but destabilized the product. We would like to avoid that. 

A good first step might be a design document that outlines our current solution. If we find blockers, then we can stop early without wasting too much time.

Arrow Flight SQL popularity increases, so it might be a great addition for Trino.

**Comment by RoeyoOgen** [comment]
@devozerov any update on the above?

**Comment by heman026** [comment]
Any update guys.

**Comment by utkarshkoppikar7** [comment]
Any update people?
